src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt

```diff
@@ -63,13 +63,19 @@ class DecodeProcessor(
                 val decodedGroup = codec.decode(messageGroup)
 
                 if (decodedGroup.messagesCount < messageGroup.messagesCount) {
-                    parentEventId.onEvent("Decoded message group contains less messages (${decodedGroup.messagesCount}) than encoded one (${messageGroup.messagesCount})")
+                    parentEventIds.forEachEvent("Decoded message group contains less messages (${decodedGroup.messagesCount}) than encoded one (${messageGroup.messagesCount})")
                 }
 
                 messageBatch.addGroups(decodedGroup)
             } catch (throwable: Throwable) {
-                parentEventId.onErrorEvent("Failed to decode message group", messageGroup.messageIds, throwable)
-                messageBatch.addGroups(messageGroup.toErrorMessageGroup(throwable, protocols))
+                val header = "Failed to decode message group"
+                val eventIds = parentEventIds.associateWith {
+                    EventUtils.toEventID(it.onErrorEvent(header, messageGroup.messageIds, throwable).id)!!
```

Comment:
Please, don't use `!!` in code. You can use `checkNotNull` with meaningful message instead